### DISCUSSION TITLE
How to make a custom comparison in pytest style?

### DISCUSSION THREAD

**Discussion by heckad** [pytest-dev/pytest#8331] [Q&A]
Hi everybody.

I have two python builtin object instance without `__eq__` method and I want to write a custom assert equal for this case. In the documentation, I found pytest_assertrepr_compare hook which does what I want but they call only when test return False. Does pytest have a hook which called instead of build-in `__eq__` method?  For example something like

foo.py
```python
# It only has the build in __cmp__ method!
class Foo:
    def __init__(self, val):
        self.val = val
```

conftest.py
```python
# If exists hoock
def pytest_asserttest(op, left, right):
   if isinstance(left, Foo) and isinstance(right, Foo) and op == "==":
      # do some cmp work
   # If return None then do build in __cmp__
```

test.py
```python

def test_something()
    f1 = Foo(1)
    f2 = Foo(2)
    assert f1 == f2  # should be cal pytest_asserttest
```

**Comment by symonk** [comment]
I am confused by your use case, it seems overly complex.  Would you not consider implementing dunder __eq__ (and Dunder __hash__) etc in your object explicitly and then use simple asserting to check the fact, passing the logic to the hook seems weird to me, not to mention non reusable

**Reply by heckad** [reply]
Unfortunately, I am getting an object from outside and when I try to do 
```
type(first).__eq__ = castom_eq
```
I got 
```
TypeError: can't set attributes of built-in/extension type 'lxml.etree._Element'
```
In `unittest` I can override `_baseAssertEqual` method, but in pytest I didn't find an idiomatic solution for this.

**Reply by RonnyPfannschmidt** [reply]
https://docs.pytest.org/en/latest/assert.html#defining-your-own-explanation-for-failed-assertions

**Reply by RonnyPfannschmidt** [reply]
oh wait, i misunderstood op

basically the way to do this is to intorduce a decorator object pattern, which translates to a wrapper object that implements eq as intendd

**Reply by heckad** [reply]
Why not add assert customization if pytest rewrites it anyway? And it would be generally cool to add hooks with a filter by operation and types of arguments.

**Reply by RonnyPfannschmidt** [reply]
Because magic behaviour hacks make debugging hell

**Reply by nicoddemus** [reply]
For the sake of discussion, currently pytest rewrites:

```python
assert <expr>
```

Into:

```python
if not <expr>:
    # magic to generate a nice message, by breaking <expr> into its components
    raise AssertionError(msg)
```

Which is a pretty harmless translation, because we don't really change any semantics.

To support the proposal here, it would need to break the expression even further in order to call hooks and decide to fail the assertion. At first glance seems complex and a potential source for bugs, but I might be wrong.

**Reply by heckad** [reply]
Pytest has the `pytest_assertrepr_compare` wich do that I want but they call after eval `<expr>` and used only to show a reason for a failed test.

**Reply by RonnyPfannschmidt** [reply]
I believe it's fine to introduce a assertrepr_valur hook for only the result valve

It's not fin to reuse the compare hook

**Reply by nicoddemus** [reply]
> Pytest has the pytest_assertrepr_compare wich do that I want but they call after eval <expr> and used only to show a reason for a failed test.

Indeed, but my point was that we don't really try to do anything _until_ the expression is `False`, at which point we would get an assertion error anyway. AFAIU to include a new hook we would need to break the expression first, because we want to overwrite the original value of `<expr>`.

I don't like this idea, because this seemingly innocent change might yield different results now:

```python
assert <expr>
```

To:

```python
if not <expr>:
    # do something else
    assert False
```

With the hook in place, both examples will behave very differently: in the first example `<expr>` will be `True` (it will call the hook), but will not call the hook in the second example and will evaluate to `False`.

Seems like we are trying to abuse the assertion rewrite to inject new logic into things; injecting new logic via mocking is not be a problem in general for "normal" Python objects, but is indeed a problem for C extension code, but I don't think we should use the assertion rewriting mechanism to circumvent that.

So far I'm :-1: on the idea.
